### DISCUSSION TITLE
Unable to connect with field in practice mode

### DISCUSSION THREAD

**Discussion by TerryGrimley** [gizmo-platform/gizmo#8] [Q&A] [answered]
Using gizmo 0.0.11-rc6 - followed the instructions for configuring the firmware - was never asked for "use Avahi" after telling it no to drivers console.  built the firmware, installed it on the system processor.  after running hello world and spin a wheel examples on the use processor, tried the Drive a robot example.  ran the gizmo field practice with my team number.  The joystick is successfully bound.  I do not have any control function. is there a better guide to the diagnostic lights than the description in the embedded documents? from top to bottom my lights are yellow white blue green red orangish red off red  - the light on the system processor is solid green and the user proc light is flashing quickly (like I would expect based on the code)

**Comment by the-maldridge** [comment]
Prior to the release of v0.0.11-rc6, mDNS was made an always-available fallback option, which is why you were not prompted.  The documentation has now been updated to reflect this change.

The page that described the lights appears not to have made the v0.0.11 release cut, so here is the description for your reference:

  * Top LED - Network status.  This LED will change red/yellow/green depending on the nature of the wireless connection.  Red means it can't find your control computer or driver station at all, yellow means its connected but isn't yet receiving information, and green means a connection has been established and the control data is being received reliably.
  
  * Middle LED - Field Position Indicator.  After initial connectivity has been established (yellow on the network status indicator) the field position indicator will start cycling through a color wheel while it waits for control information to be received.  The colors of the wheel have no information, only that the wheel is cycling.  Once a field position has been received, the light will blink the number and color of the field.  For example if you were positioned on the blue quadrant of the second field, you would see the light blink twice every second in a blue color.  Practice mode is indicated by a white blink once per second.

  * Bottom LED - Battery Voltage Indicator.  Ranged for a 7.2v NiCad battery, this LED will cycle from green to red as the battery voltage lowers.  It is normal for the LED to show a "more dead" color when the battery is under load, the state that matters is when no motors are running and no servos are holding significant load.  Unfortunately due to a hardware defect not caught fabrication cycle for v00.r4b (version numbers are in the upper left hand corner of the board), the battery indicator is non-functional on this specific hardware revision.

After the initial flash with a firmware version that supports driver station pairing, which the firmware version that's bundled with v0.0.11 does not, the network and field LEDs will rapidly blink purple indicating that no pairing information is available.  Once pairing has been attempted, the same 2 LEDs will blink either red, yellow, or purple signifying either a format error in the config file, a syntactic error in the settings provided, or no file at all, respectively.

Based on your description of the lights on your board, it sounds like it has successfully made contact with your wireless network, but is unable to receive data on the control channel.  If you are using Windows to host the practice mode field, can you verify that Windows Firewall is allowing communication on port number 1883 to the Gizmo.exe program?

**Reply by TerryGrimley** [reply]
This is windows 11.  The firewall is set to allow gizmo.exe to use either TCP or UDP on all ports.  For grins, I added all protocols on all ports for gizmo.exe but that had no effect. The comment that the firmware currently does not support the driver's station explains why that doesn't work either.  Is there a "verbose" mode for the system firmware that might provide feedback using serial communication?  Seems like that would also be handy for the gizmo software - like feedback that showed the variable values for the joystick.  Appreciate the help, I'm just not clear what to do next.

**Reply by the-maldridge** [reply]
There is a verbose mode, unfortunately it was added after v0.0.11 shipped.  If you have a driver station, you *can* download and flash the firmware separately.  The newer firmware version will work with the v0.0.11 driver station build.  If you want to do that, the newer version can be fetched directly from the [firmware repo releases](https://github.com/gizmo-platform/firmware/releases/tag/v0.0.1).  If you connect using a serial console the system processor will spit back messages on key events at `9600,8,n,1` (default for most modern consoles).  That's probably the direction I'd recommend for troubleshooting since it provides more information as to what the system processor is doing and you can tell based on the most recent event if its gotten stuck during the connection workflow.

Its also worth noting (as I don't think this is adequately documented for the driver station anywhere yet) that if you have anything plugged into the driver station's ethernet port it will prevent the Gizmo from connecting to the driver's station wirelessly.

**Reply by TerryGrimley** [reply]
Not much luck - downloaded and installed the version linked.  copied over the Json file.  connected to watch the status info on the com port.  seems to be stuck in the gizmo_net_wifi_start.   I'm using the dirvers station - only have a batttery plugged in for power and the joystick plugged in the USB port below the power connection.
This is the configuration file that was created using the gizmo software:
{"Team":99,"UseDriverStation":true,"UseExtNet":false,"NetSSID":"fa0ab20e33a34e5585c5ddd6a0c48305","NetPSK":"42cfa82c0d99473ca7bc0a378b869258","ServerIP":"gizmo-ds"}

**Reply by the-maldridge** [reply]
Hmm, that file looks correct, and loading it onto a driver's station at my end I don't get any errors.  One thing you could try at this stage would be to try to join your phone as the other end of the driver-station to gizmo link.  I prepared this QR code using your config file that you can scan with your phone or tablet to try and connect:

![image](https://github.com/gizmo-platform/gizmo/assets/2321814/0ba1410d-2e1c-49e6-a091-977c89fd06fd)

the fact that the gizmo is stuck in `GIZMO_NET_WIFI_START` makes me think that either the network that is up isn't quite right, or the Gizmo config is somehow out of sync with the driver station.

**Reply by TerryGrimley** [reply]
Reloaded the driver's station and the json file.  I expect the issue was that they json file was generated on one computer and loaded to the drivers station, then regenerated (with the same answers) on another computer and loaded into the system processor (?).  I assume that destroys the network configuration parameters (if that's the case, might be a good head's up for the documentation ).   I was able to connect with my phone after redoing the driver's station, and now get farther with the gizmo, but still not all the way to functioning.
messages are below and I'm caught in a disconnect/connect cycle - I assume the watchdog is hitting this every 15 seconds.  I assume this means it's connecting to the driver's station but no information is flowing (or being recognized)?
GIZMO_LOAD_ATTEMPT
GIZMO_LOAD_JSON_OK
GIZMO_LOAD_CFG_OK
GIZMO_NET_WIFI_START
GIZMO_NET_WIFI_START
GIZMO_NET_WIFI_RUN
GIZMO_FMS_DISCOVERED_HOSTNAME
GIZMO_MQTT_CONNECT_OK
GIZMO_MQTT_DISCONNECT
GIZMO_MQTT_CONNECT_OK

**Reply by the-maldridge** [reply]
Ah, yeah I will make a note that we need to document that better.  The same answers will result in different security keys since the network parameters are generated from random sources at the time the wizard is run.  The driver station is actually capable of completing the parameter load process (referred to in the docs as [binding](https://gizmo-platform.github.io/startup/firmware.html#bind-the-gizmo-to-a-drivers-station)).  You may find that workflow to be more convenient than using a second computer, and it is the workflow we would expect most end-users to follow rather than pulling out a second computer (the config-server isn't compiled in on Windows).

The event stream output you posted above does seem to show its connecting, but the fact that its getting an MQTT disconnect does mean that the watchdog is realizing it hasn't received any configuration information in the initial 15s timeout after connecting.  I'm trying to think of a good way to debug this that doesn't involve connecting either a full computer to the wireless link to observe what's going on or having to hook up the driver's station to a monitor.  I'll think on this and add another reply later this evening.

**Reply by the-maldridge** [reply]
Right now the only thought I've got is to start introspecting the state on the driver station.  If you have a mini-HDMI to full size HDMI adapter, it would be good to plug the driver station into a monitor.  It dumps its logs to the screen and any serious errors should be visible there.

**Reply by TerryGrimley** [reply]
I appreciate all of the support here - I fully expect that it's something that I missed and/or just did wrong.  I don't have a mini-hdmi connector, but I'll get one. (Apparently, I have a micro hdmi if that is what was used on the PI-4.)

**Reply by TerryGrimley** [reply]
I booted with my original micro-sd and saw at least one disturbing message, so I used a second imaged card with the proper gson file and restarted.  I still don't get use of the joystick but the boot message from the drivers station looks cleaner to me.
![bad boot](https://github.com/gizmo-platform/gizmo/assets/166930489/04caf15d-bbf2-4f9a-9c32-38ef037434d4)
![good boot](https://github.com/gizmo-platform/gizmo/assets/166930489/980c86ad-4c0f-4bcd-8367-9dbed6bd4cdb)

**Reply by the-maldridge** [reply]
My apologies for the slow replies here, I was called away to other matters unexpectedly.

At this point, I think the next step to try and work out what is different from your setup to mine is to log in and for you to try and run the practice mode commands directly.

Fortunately, in the edition of the gizmo driver station you're using, the password is static, you can log in as `root` with the password `voidlinux`.  From there type `sv down gizmo-practice` to stop the embedded practice field and you can run the gizmo commands as you would on any other computer.  The `gsscfg.json` can be found at `/boot/gsscfg.json` when logged in.

For context, we're way past where anything should have broken here, but I do think its important for us to root cause this to find either the defect in the docs, processes that we advised, or the unlikely event of a defect in the code itself.  I really do appreciate your time and effort in helping us improve the platform.

**Reply by TerryGrimley** [reply]
I used the commands provided to login to the driver station, stop the gizmo practice server and then started it manually.  Nothing surprising/alarming when I did that.  The status messages showed the system recognized the joystick. When I connected a USB cable between the system processor and the drivers station and restarted the system proc, I could see the messages on the drivers station indicating that a configuration was requested and provided.  After restarting the gizmo, the middle light continued to cycle.  Just to verify something (I tried it both ways), but is it correct that the practice jumper is not needed/part of this approach?  Other than a label on an image of the circuit board, I didn't find anything indicating the jumper being open/closed for any particular operation, but I'd hate to be chasing this and find out it's just because I failed to put a jumper in place (again, tried both ways and saw no difference).  Is there a simple way to manually consume the json messages using a browser? Same thing for manually sending packets?

**Comment by TerryGrimley** [comment, marked as answer]
So... the update that was published a couple of days ago now allows my gizmo to connect to the driver's station (I think) - the top light is a happy green, the middle light is flashing blue(?) a couple of times per second (flash-pause equal, not flash-flash-pause).  So yay for that, I think. Initially, I got no Joystick values (using the sample code), but my last try was to switch USB ports on the driver's station for the Joystick and try again, and that worked.  Is it supposed to matter which USB port you use on the driver's station?
